src/runtime_src/core/pcie/emulation/common_em/utility.cxx

```diff
@@ -0,0 +1,89 @@
+#include "utility.h"
+#include <unistd.h>
+#include <signal.h>
+#include <sstream>
+#include <wait.h>
+
+#include <cstdio>
+#include <iostream>
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <array>
+#include <fstream>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include <stdlib.h>
+#include <dirent.h>
+#include <sys/types.h>
+
+
+std::string cUtility::launch_bash(const char* fullcmd)
+{
+    std::array<char, 128> buffer;
+    std::string result;
+    std::shared_ptr<FILE> pipe(popen(fullcmd, "r"), pclose);
+
+    if (!pipe)
+    {
+        throw std::runtime_error("popen() failed!");
+    }
+
+    while (!feof(pipe.get()))
+    {
+        if (fgets(buffer.data(), 128, pipe.get()) != nullptr)
+        {
+            result += buffer.data();
+        }
+    }
+
+    return result;
+}
+
+
+pid_t cUtility::proc_find(const char* name) 
+{
+    DIR* dir;
+    struct dirent* ent;
+    char* endptr;
+    char buf[512];
+
+    if (!(dir = opendir("/proc"))) {
+        std::cout<<"can't open /proc";
+        return -1;
+    }
+
+    while((ent = readdir(dir)) != NULL) {
+        /* if endptr is not a null character, the directory is not
+         * entirely numeric, so ignore it */
+        long lpid = strtol(ent->d_name, &endptr, 10);
+        if (*endptr != '\0') {
+            continue;
+        }
+
+        /* try to open the cmdline file */
+        snprintf(buf, sizeof(buf), "/proc/%ld/cmdline", lpid);
+        FILE* fp = fopen(buf, "r");
+
+        if (fp) {
+            if (fgets(buf, sizeof(buf), fp) != NULL) {
+                /* check the first token in the file, the program name */
+                char* first = strtok(buf, " ");
+                if (std::string(first).find(name) != std::string::npos) {
+                    fclose(fp);
+                    closedir(dir);
+                    return (pid_t)lpid;
+                }
+            }
+            fclose(fp);
+        }
```

Comment:
fopen is not C+ and what if in-between code throws an exception.
NULL is not C++.
What if strtok is nullptr?